pkg/base/registry/nacos/registry.go

```diff
@@ -0,0 +1,173 @@
+	"github.com/transaction-wg/seata-golang/pkg/tc/config"
+	"github.com/transaction-wg/seata-golang/pkg/util/log"
+)
+
+func init() {
+	extension.SetRegistry(constant.NACOS_KEY, newNacosRegistry)
+}
+
+type nacosRegistry struct {
+	namingClient naming_client.INamingClient
+}
+type nacosEventListener struct {
+}
+
+func (nr *nacosEventListener) OnEvent(service []*registry.Service) error {
+	data, err := json.Marshal(service)
+	log.Info("servie info change：" + string(data))
+	return err
+}
+func (nr *nacosRegistry) Register(addr *registry.Address) error {
+	registryConfig := config.GetRegistryConfig()
+
+	param := createRegisterParam(registryConfig, addr)
+	isRegistry, err := nr.namingClient.RegisterInstance(param)
+	if err != nil {
+		return err
+	}
+	if !isRegistry {
+		return errors.Errorf("registry [" + registryConfig.NacosConfig.Application + "] to  nacos failed")
+	}
+	return nil
+}
+
+//创建服务注册信息
+func createRegisterParam(registryConfig config.RegistryConfig, addr *registry.Address) vo.RegisterInstanceParam {
+	serviceName := registryConfig.NacosConfig.Application
+	params := make(map[string]string)
+
+	instance := vo.RegisterInstanceParam{
+		Ip:          addr.IP,
+		Port:        addr.Port,
+		Metadata:    params,
+		Weight:      1,
+		Enable:      true,
+		Healthy:     true,
+		Ephemeral:   true,
+		ServiceName: serviceName,
+		ClusterName: registryConfig.NacosConfig.Cluster, // default value is DEFAULT
+		GroupName:   registryConfig.NacosConfig.Group,   // default value is DEFAULT_GROUP
+	}
+	return instance
+}
+
+func (nr *nacosRegistry) UnRegister(addr *registry.Address) error {
+	return nil
+}
+
+//noinspection ALL
+func (nr *nacosRegistry) Lookup() ([]string, error) {
+	registryConfig := clientConfig.GetRegistryConfig()
+	clusterName := registryConfig.NacosConfig.Cluster
+	instances, err := nr.namingClient.SelectInstances(vo.SelectInstancesParam{
+		ServiceName: registryConfig.NacosConfig.Application,
+		GroupName:   registryConfig.NacosConfig.Group, // default value is DEFAULT_GROUP
+		Clusters:    []string{clusterName},            // default value is DEFAULT
+		HealthyOnly: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	addrs := make([]string, 0)
+	for _, instance := range instances {
+		addrs = append(addrs, instance.Ip+":"+strconv.FormatUint(instance.Port, 10))
+	}
+	//订阅服务
+	nr.Subscribe(&nacosEventListener{})
+	return addrs, nil
+}
+func (nr *nacosRegistry) Subscribe(notifyListener registry.EventListener) error {
+	registryConfig := clientConfig.GetRegistryConfig()
+	clusterName := registryConfig.NacosConfig.Cluster
+	err := nr.namingClient.Subscribe(&vo.SubscribeParam{
+		ServiceName: registryConfig.NacosConfig.Application,
+		GroupName:   registryConfig.NacosConfig.Group, // default value is DEFAULT_GROUP
+		Clusters:    []string{clusterName},            // default value is DEFAULT
+		SubscribeCallback: func(services []model.SubscribeService, err error) {
+			serviceList := make([]*registry.Service, 0)
```

Comment:
This new slice capacity can be specified, otherwise there will be heap allocation for below append.

```
serviceList := make([]*registry.Service, 0, len(services))
```